File: include/madara/knowledge_engine/containers/Integer.cpp
#include "Integer.h"


Madara::Knowledge_Engine::Containers::Integer::Integer (
  const Knowledge_Update_Settings & settings)
: context_ (0), settings_ (settings) 
{
}
  
Madara::Knowledge_Engine::Containers::Integer::Integer (
  const std::string & name,
  Knowledge_Base & knowledge,
  const Knowledge_Update_Settings & settings)
: context_ (&(knowledge.get_context ())), name_ (name), settings_ (settings) 
{
  variable_ = knowledge.get_ref (name);
}
 
Madara::Knowledge_Engine::Containers::Integer::Integer (
  const std::string & name,
  Variables & knowledge,
  const Knowledge_Update_Settings & settings)
: context_ (knowledge.get_context ()), name_ (name), settings_ (settings) 
{
  variable_ = knowledge.get_ref (name);
}
 
Madara::Knowledge_Engine::Containers::Integer::Integer (
  const std::string & name,
  Knowledge_Base & knowledge,
  type value,
  const Knowledge_Update_Settings & settings)
: context_ (&(knowledge.get_context ())), name_ (name), settings_ (settings) 
{
  variable_ = knowledge.get_ref (name);
  context_->set (variable_, value, settings);
}
 
Madara::Knowledge_Engine::Containers::Integer::Integer (
  const std::string & name,
  Variables & knowledge,
  type value,
  const Knowledge_Update_Settings & settings)
: context_ (knowledge.get_context ()), name_ (name), settings_ (settings) 
{
  variable_ = knowledge.get_ref (name);
  context_->set (variable_, value, settings);
}


Madara::Knowledge_Engine::Containers::Integer::Integer (const Integer & rhs)
  : context_ (rhs.context_),
    name_ (rhs.name_),
    variable_ (rhs.variable_),
    settings_ (rhs.settings_)
{

}


Madara::Knowledge_Engine::Containers::Integer::~Integer ()
{

}

void
Madara::Knowledge_Engine::Containers::Integer::exchange (
  Containers::Integer & other)
{
  Guard guard (mutex_), guard2 (other.mutex_);

  type temp = *other;
  other = **this;
  *this = temp;
}

std::string
Madara::Knowledge_Engine::Containers::Integer::get_name (void)
{
  Guard guard (mutex_);
  return name_;
}

void
Madara::Knowledge_Engine::Containers::Integer::set_name (
  const std::string & var_name,
  Knowledge_Base & knowledge)
{
  context_ = &(knowledge.get_context ());
  name_ = var_name;
  variable_ = context_->get_ref (name_, settings_);
}

void
Madara::Knowledge_Engine::Containers::Integer::set_name (
  const std::string & var_name,
  Variables & knowledge)
{
  context_ = knowledge.get_context ();
  name_ = var_name;
  variable_ = context_->get_ref (name_, settings_);
}

Madara::Knowledge_Engine::Containers::Integer::type
Madara::Knowledge_Engine::Containers::Integer::operator= (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    context_->set (variable_, value, settings_);
  }

  return value;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator== (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) == value;
  }

  return false;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator != (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) != value;
  }

  return true;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator== (const Integer & value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return
      context_->get (variable_, settings_) ==
        value.context_->get (value.variable_, value.settings_);
  }

  return false;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator != (const Integer & value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return
      context_->get (variable_, settings_) !=
        value.context_->get (value.variable_, value.settings_);
  }

  return true;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator< (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) < value;
  }

  return false;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator<= (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) <= value;
  }

  return false;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator> (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) > value;
  }

  return false;
}

bool
Madara::Knowledge_Engine::Containers::Integer::operator>= (type value)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_) >= value;
  }

  return false;
}

Madara::Knowledge_Engine::Containers::Integer::type
Madara::Knowledge_Engine::Containers::Integer::operator* (void)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_).to_integer ();
  }
  else
    return 0;
}
   
double
Madara::Knowledge_Engine::Containers::Integer::to_double (void)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_).to_double ();
  }
  else
    return 0.0;
}
   
std::string
Madara::Knowledge_Engine::Containers::Integer::to_string (void)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    return context_->get (variable_, settings_).to_string ();
  }
  else
    return "";
}

void
Madara::Knowledge_Engine::Containers::Integer::set_quality (
  uint32_t quality,
  const Knowledge_Reference_Settings & settings)
{
  Guard guard (mutex_);
  
  if (context_)
  {
    context_->set_quality (name_, quality, true, settings);
  }
}

File: port/java/src/com/madara/containers/Integer.java
/*********************************************************************
 * Usage of this software requires acceptance of the MADARA CMU License,
 * which can be found at the following URL:
 *
 * https://code.google.com/p/madara/wiki/CMULicense
 *********************************************************************/
package com.madara.containers;

import com.madara.MadaraJNI;
import com.madara.KnowledgeBase;
import com.madara.Variables;

public class Integer extends MadaraJNI
{	
  private native long jni_Integer();
  private native long jni_Integer(long cptr);
  private static native void jni_freeInteger(long cptr);
  private native void jni_set(long cptr, long value);
  private native java.lang.String jni_getName(long cptr);
  private native void jni_setName(long cptr, long type, long kb, java.lang.String name);
  private native java.lang.String jni_toString(long cptr);
  private native double jni_toDouble(long cptr);
  private native long jni_toLong(long cptr);

  private boolean manageMemory = true;

  public Integer()
  {
    setCPtr(jni_Integer());
  }

  public Integer(Integer input)
  {
    setCPtr(jni_Integer(input.getCPtr()));
  }

  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @return a new java instance of the underlying pointer
   */
  public static Integer fromPointer(long cptr)
  {
    Integer ret = new Integer();
    ret.manageMemory = true;
    ret.setCPtr(cptr);
    return ret;
  }

  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @param shouldManage  if true, manage the pointer
   * @return a new java instance of the underlying pointer
   */
  public static Integer fromPointer(long cptr, boolean shouldManage)
  {
    Integer ret = new Integer();
    ret.manageMemory=shouldManage;
    ret.setCPtr(cptr);
    return ret;
  }

  /**
   * Gets the value
   *
   * @return   current value
   */
  public long get()
  {
    return jni_toLong(getCPtr());
  }

  /**
   * Gets the name of the variable
   *
   * @param  name of the variable within the context
   */
  public java.lang.String getName()
  {
    return jni_getName(getCPtr());
  }

  /**
   * Sets the value
   *
   * @param  value   new value
   */
  public void set(long value)
  {
    jni_set(getCPtr(), value);
  }

  /**
   * Sets the name and knowledge base being referred to
   *
   * @param  kb      the knowledge base that contains the name
   * @param  name    the variable name
   */
  public void setName(KnowledgeBase kb, java.lang.String name)
  {
    jni_setName(getCPtr(), 0, kb.getCPtr (), name);
  }

  /**
   * Sets the name and knowledge base being referred to
   *
   * @param  vars    the variables facade that contains the name
   * @param  name    the variable name
   */
  public void setName(Variables vars, java.lang.String name)
  {
    jni_setName(getCPtr(), 1, vars.getCPtr (), name);
  }

  /**
   * Converts the value to a double
   *
   * @return current double value
   */
  public double toDouble()
  {
    return jni_toDouble(getCPtr());
  }

  /**
   * Converts the value to a long
   *
   * @return current long value
   */
  public long toLong()
  {
    return jni_toLong(getCPtr());
  }

  /**
   * Converts the value to a string
   *
   * @return current string value
   */
  public java.lang.String toString()
  {
    return jni_toString(getCPtr());
  }

  /**
   * Deletes the C instantiation. To prevent memory leaks, this <b>must</b> be
   * called before an instance gets garbage collected
   */
  public void free()
  {
    if (manageMemory)
    {
      jni_freeInteger(getCPtr());
    }
  }
}

